lib/pure/hashes.nim

```diff
@@ -123,6 +123,33 @@ proc hiXorLo(a, b: uint64): uint64 {.inline.} =
     else:
       result = hiXorLoFallback64(a, b)
 
+when defined(js):
+  import std/jsbigints
+  import std/private/jsutils
+
+  proc noBigInt(): bool =
+    asm """`result` = typeof BigInt == 'undefined'"""
+
+  proc hiXorLoJs(a, b: JsBigInt): JsBigInt =
+    let prod = a * b
+    let mask = (big"1" shl big"64") - big"1"
+    result = (prod shr big"64") xor (prod and mask)
+
+  template hashWangYiJS(x: JsBigInt): Hash =
+    let P0 = (big"0xa0761d64" shl big"32") or big"0x78bd642f"
+    let P1 = (big"0xe7037ed1" shl big"32") or big"0xa0b428db"
+    let P58 = (big"0xeb44acca" shl big"32") or (big"0xb455d165" xor big"8")
+    let res = hiXorLoJs(hiXorLoJs(P0, x xor P1), P58)
+    let maxSafeInterger = (big"1" shl big"53") - big"1"
```

Comment:
in future work we need to expose `Number.isSafeInteger`, `MAX_SAFE_INTEGER` + friends (but not in this PR)

Reply:
=> https://github.com/timotheecour/Nim/issues/561